packages/styled-components/src/models/StyledComponent.js

```diff
@@ -298,6 +298,31 @@ export default function createStyledComponent(target: Target, options: Object, r
   // $FlowFixMe
   WrappedStyledComponent.target = isTargetStyledComp ? target.target : target;
 
+  // fold defaultProps
+  // $FlowFixMe
+  WrappedStyledComponent._defaultProps = EMPTY_OBJECT;
+  // $FlowFixMe
+  Object.defineProperty(WrappedStyledComponent, 'defaultProps', {
```

Comment:
Why do we need to use a setter here? Doing the merge once at construction time should be sufficient I think...

Reply:
Hey! Thanks for reviewing. 
That's what I did with my first implementation, but I had to merge twice
```js
const A = styled.div``
A.defaultProps = {}

const B = styled(A)`` // A's defaultProps at construction time
B.defaultProps = {}

<B />  // B's defaultProps at render time
```

I would also have to modify this: https://github.com/styled-components/styled-components/blob/e6f1b85726512cde163dd5fcdd7fb0f2439d4a09/packages/styled-components/src/models/StyledComponent.js#L125

I can totally code it without a setter, but I thought it was more complex to read.

Reply:
> I would also have to modify this

Why? If you do the merge once at construction time (during `createStyledComponent`), the new defaultProps will be correct for when `determineTheme` is used in that component. Just set defaultProps on `WrappedStyledComponent`.

Reply:
Yeah sorry the problem was elsewhere, here are the problems I was facing: 

1) I don't think it's possible to fold directly on `WrappedStyledComponent.defaultProps`.
  ```js
  // Folding defaultProps' target to WrappedStyledComponent.defaultProps at construction time
  const A = styled(X)``
  A.defaultProps = { color: 'red' }
  // WrappedStyledComponent.defaultProps is now  { color: 'red' } and we lost everything.
  // WrappedStyledComponent.defaultProps is A.defaultProps
  ```
Easily fixable by using another static property to preserve the context (that I'm going to call _defaultProps for the example)

2) Like I wrote there https://github.com/styled-components/styled-components/pull/2260#discussion_r267719350 I was merging at construction time and at render time. 

3) The main difficulty appears at render time, https://github.com/styled-components/styled-components/blob/e6f1b85726512cde163dd5fcdd7fb0f2439d4a09/packages/styled-components/src/models/StyledComponent.js#L279

`defaultProps` are already applied by React, we have access to them (but are not merged yet), _defaultProps (merged except the last defaultProps), props (is a mix of defaultProps and props), and we still need to merge defaultProps (including style) and reapply them. 

Example:
```js
const A = styled.div``
A.defaultProps = { style: { color: 'red' } }
<A /> // expected: color: red

const B = styled(A)``
B.defaultProps = { style: { backgroundColor: 'red' } }
// Case B1
<B /> // expected: color: red and backgroundColor: red
// Case B2
<B style={{ color: blue }} />  // expected: color: blue
````
```js
// At render time
B1 => B1._defaultProps == { style: { color: 'red' } }
   => B1.defaultProps == { style: { backgroundColor: 'red' } }
   => B1.props == { style: { backgroundColor: 'red' } }
result expected => { style: { backgroundColor: 'red', color: 'red' } }

B2 => B2._defaultProps == { style: { color: 'red' } }
   => B2.defaultProps == { style: { backgroundColor: 'red' } }
   => B2.props == { style: { color: 'blue' } }
result expected => { style: { color: 'blue' } }
```

By using a setter and merging defaultProps before the rendering, everything works as expected. Maybe I missed something...